### PR TITLE
Restrict embedding projection to knn based kernels

### LINKED ISSUE #668: Embedding projection for non-knn based kernels
- [ ] Additional function parameters, change functionality or change defaults?
- [ ] New estimator in `cr.tl.estimators`?
- [ ] New kernel in `cr.tl.kernels`?
- [ ] New gene trend model in `cr.ul.models`?
- [ ] New plotting function in `cr.pl`?
- [x] Other?

## Description

#665 fixes the embedding projection for knn based kernels. However, the fix restricts the method to only these kernels. As already discussed in #603, the (original) embedding projection did not work correctly for `WOTKernel`. As outlined [here](https://github.com/theislab/cellrank/issues/603#issuecomment-877067538) in #603, using the originally implemented projection on the on reprogramming data from Schiebinger et al., Cell 2019 leads to 

original

Omitting the random uniform correction term leads to improved results:

wo_random_uniform

This is also true when weighing by inverse connectivities:

w_weighted_conns

As discussed [here](https://github.com/theislab/cellrank/issues/603#issuecomment-877067538), results look not as promising for the `PseudoTimeKernel`.

### PR DESCRIPTION
**Pull request by WeilerP**
**IMPORTANT: Please search among the [Pull requests](../pulls) before creating one.**

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)

## Description

Restricts computation of embedding projection to kNN based kernels.

## How has this been tested?

`TestOTKernel` and `TestWOTKernel` are extended by one unit test each to verify that expected error is thrown.

## Closes

Closes #668.

### CODE REVIEW

**Comment by WeilerP** [comment]
I'll rebase once #734 has been solved.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: cellrank/tl/kernels/_base_kernel.py [hunk 1]
```diff
@@ -299,10 +299,10 @@ def compute_projection(
             )
 
         kernels_w_conn = [kernel for kernel in self.kernels if kernel._conn is not None]
```
[the remark this answers is not in the document]
Comment (WeilerP) [reply cellrank/tl/kernels/_base_kernel.py:301]: Oh yes, you're right!

**Review by michalk8 — APPROVED** [review]

**Comment by michalk8** [comment]
@WeilerP I'd use `{kernel!r}` since `__repr__` doesn't print out the params, this should fix the `match=...` in those tests. Afterwards, it can be merged.

**Comment by Marius1311** [comment]
Really nice! Glad to finally close this issue!

**Comment by michalk8** [comment]
@WeilerP think this can be merged, right?

**Comment by WeilerP** [comment]
> @WeilerP think this can be merged, right?

Yes!

**Pull request merged**